### PR TITLE
Refactor large parser modules into smaller components under 400 lines (#223)

### LINKED ISSUE #223: [High] Oversized modules violate local maintainability guideline (400-line cap)
## Description

Several modules exceed the local maintainability guideline of 400 lines, making them harder to understand, test, and maintain.

## Affected Modules

The following files violate the 400-line cap:

- `src/parser/ast/rule.rs`
- `src/parser/ast/expr.rs`
- `src/parser/expression/pratt.rs`
- `src/parser/tests/expression.rs`

## Impact

Oversized modules:
- Reduce code readability and comprehension
- Make testing and maintenance harder
- Increase cognitive load for contributors
- Violate established project maintainability guidelines

## Actionable Fixes

**`src/parser/ast/rule.rs`**
- Move rule-term classification logic into a separate module

**`src/parser/ast/expr.rs`**
- Extract S-expression formatting into a dedicated module

**`src/parser/expression/pratt.rs`**
- Split postfix/diff/delay handling into separate focused modules

**`src/parser/tests/expression.rs`**
- Move giant test case tables into data fixtures

## Recommendation

Split these modules by responsibility, following the single-responsibility principle. Each resulting module should focus on a specific concern and stay well under the 400-line guideline.

## References

- PR: https://github.com/leynos/ddlint/pull/212
- Requested by `@leynos`

## Requested by

`@leynos`

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Refactors large parser modules into smaller, focused components to improve maintainability and testability. No user-facing behavior changes.

### Core Refactor
- Introduced dedicated S-expression formatting for Expr via a new module: src/parser/ast/expr/sexpr.rs and wired it into the Expr API (Expr::to_sexpr moved from the main file).
- Moved rule-body expression classification logic into a dedicated module: src/parser/ast/rule/classification.rs and updated rule.rs to delegate classification to it.
- Extracted Pratt postfix handling into separate modules:
  - src/parser/expression/pratt/delay.rs
  - src/parser/expression/pratt/diff.rs
  - src/parser/expression/pratt/postfix.rs
  - Updated src/parser/expression/pratt.rs to wire in the new modules (mod delay; mod diff; mod postfix).
- Adjusted module layout and imports to reflect the new structure, reducing file sizes and coupling in core parser files.

### Tests and Fixtures
- Reworked tests to rely on a structured fixtures system for expression parsing:
  - Added new test fixtures module: src/parser/tests/expression/fixtures.rs
  - Split valid expression fixtures into submodules (basic, collections, control_flow, postfix, structured) and wired them into a single expression_cases() entry point
  - Added error fixtures at: src/parser/tests/expression/fixtures/errors.rs
  - Added postfix error fixtures at: src/parser/tests/expression/fixtures/valid/postfix.rs and related fixtures modules
- Introduced a dedicated sexpr rendering helper for tests (via the new sexpr.rs) and updated tests to use to_sexpr() for compact comparison of parsed structures.

### Miscellaneous
- Minor cleanup to remove inline to_sexpr implementation from Expr and rely on the new sexpr module.
- Code organization improvements align with the 400-line module guideline mentioned in the task, while preserving functionality.

## How to test
- Run: cargo test
- Specifically validate that expression parsing tests pass with the new fixtures layout and that S-expression representations render as expected via Expr::to_sexpr().
- Ensure no public API changes in user-facing behavior; changes are internal restructuring and test shims.

## Notes
- This refactor improves maintainability and testability of parser components by modularizing: S-expression formatting, rule classification, and Pratt postfix handling.
- If any downstream tooling relies on file locations, please update accordingly; public behavior remains unchanged.

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---

ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/4e7f979f-30bc-45b5-844c-4e890802856c

📝 Closes #223

## Summary by Sourcery

Refactor parser internals and tests into smaller, focused modules while preserving existing parsing behavior.

Enhancements:
- Extract rule body classification logic into a dedicated module and delegate from the main rule AST type.
- Split Pratt parser postfix, diff, and delay handling into separate helper modules to reduce pratt.rs size and coupling.
- Move S-expression formatting for expressions into its own module to decouple test-oriented rendering from the core AST definition.
- Introduce a shared expression test fixtures module with structured valid and error cases to simplify and consolidate parser tests.

Tests:
- Rewrite expression parser tests to consume structured fixtures for literals, control flow, postfix operations, and error conditions, improving reuse and coverage organization.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/parser/tests/expression/fixtures/errors.rs](https://github.com/leynos/ddlint/pull/259/files/8802babd023dde765953f618c4e7c19ce6a095bb#diff-1786604f926e7e004d13dd6c5915774da79b0e4bbb563fddf6148a191668cbd2)

Comment on lines +217 to +232

~~~rust
fn collection_error_cases() -> Vec<CountedErrorCase> {
    vec![
        CountedErrorCase {
            src: "[1, 2",
            min_errs: 1,
        },
        CountedErrorCase {
            src: "{a: 1",
            min_errs: 1,
        },
        CountedErrorCase {
            src: "{a, b}",
            min_errs: 1,
        },
    ]
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69278/delta?repo-id=1160463&review-id=259&biomarker=Code+Duplication&filename=src%2Fparser%2Ftests%2Fexpression%2Ffixtures%2Ferrors.rs&method=)
The module contains 4 functions with similar structure: collection_error_cases,control_flow_error_cases,operator_error_cases,postfix_counted_error_cases

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/parser/expression/pratt.rs](https://github.com/leynos/ddlint/pull/259/files/8802babd023dde765953f618c4e7c19ce6a095bb#diff-04b655afdc1da4cba37c25876da245599553ec045859ccfaaeb6b3c9acce3e04)

Comment on file

~~~rust
use chumsky::error::Simple;

use crate::parser::ast::{Expr, StringLiteral};
use crate::parser::span_utils::parse_u32_decimal;
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69278/delta?repo-id=1160463&review-id=259&biomarker=Code+Duplication&filename=src%2Fparser%2Fexpression%2Fpratt.rs&method=)
The module contains 2 functions with similar structure: Pratt.with_struct_literal_activation,Pratt.with_struct_literals_suspended

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/parser/tests/expression/fixtures/errors.rs: 

### What lead to degradation?
The module contains 4 functions with similar structure: collection_error_cases,control_flow_error_cases,operator_error_cases,postfix_counted_error_cases

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | PR introduces new postfix parsing modules (delay.rs, diff.rs) but test fixtures completely omit passing test cases for diff markers and delay operators, making tests vacuous. | Add test cases for diff marker (e.g. `e'`) and delay parsing (e.g. `x-<5>`) to valid/postfix.rs. Apply whitespace-trimming fix to classification.rs line 61. Add test in body\_terms.rs for whitespace-only values. |
| Module-Level Documentation | ❌ Error | Three Pratt parser modules (delay.rs, diff.rs, postfix.rs) lack sufficient docstrings explaining utility and function beyond stating purpose. | Expand docstrings in delay.rs, diff.rs, and postfix.rs to clarify utility, function, and rationale following the pattern established in sexpr.rs and classification.rs. |
| Unit Architecture | ❌ Error | Diff/delay postfix parsing modules lack dedicated unit test fixtures, violating the requirement that all new functionality must be guarded by tests. | Add test cases for diff and delay postfix expressions to src/parser/tests/expression/fixtures/valid/postfix.rs covering both foo'(bar) and foo-<5> forms. |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 50.00% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Developer Documentation | ⚠️ Warning | Pull request refactors parser modules below 400-line guideline but extracted modules lack required developer guide documentation. | Create `docs/developers-guide.md` documenting new module structure, responsibilities (sexpr formatting, rule classification, postfix/delay/diff parsing), and architectural boundaries. |
| Testing (Unit And Behavioural) | ⚠️ Warning | PR fails critical testing requirements: missing test cases for diff and delay postfix operations, missing .trim() fix for whitespace-only assignment check, and new extracted modules lack dedicated unit tests. | Add test cases for diff (`'`) and delay (`-<number>`) postfix forms to `postfix.rs`; apply `.trim()` to line 61 in `classification.rs`; add unit tests for sexpr, classification, delay, diff, and postfix modules. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | PR introduces text-based, structured S-expression output lacking snapshot tests and omits diff/delay fixtures from the refactored fixture file despite explicit review request. | Add diff and delay test cases to src/parser/tests/expression/fixtures/valid/postfix.rs. Implement snapshot testing for sexpr.rs module. Add focused unit tests to new modules rather than relying solely on integration-level coverage. |

**Comment by leynos** [comment]
@coderabbitai please create new GitHub issues for the following:

## Out-of-scope: GitHub issues required

### Docstring Coverage (⚠️ Warning)

Issue `#229` ("Documentation gaps: missing module comments and inconsistent API doc sections") already tracks this. No new issue is needed — link the CodeScene warning to `#229`.

### Developer Documentation (⚠️ Warning) and Snapshot Testing (⚠️ Warning)

These two are out of scope for this refactoring PR. Shall I open GitHub issues for:

1. **Create `docs/developers-guide.md`** documenting the new module structure, responsibilities, and architectural boundaries introduced by this PR.
2. **Introduce snapshot testing for `sexpr.rs`** (e.g. using `insta`) to guard the S-expression rendering output against regression.

File: src/parser/tests/expression/fixtures/valid/postfix.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        ExpressionCase {
+            src: "e[1,0]",
+            expected: bit_slice(var("e"), lit_num("1"), lit_num("0")),
+        },
+        ExpressionCase {
+            src: "t.0",
+            expected: tuple_index(var("t"), "0"),
+        },
+    ]
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/parser/tests/expression/fixtures/valid/postfix.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/parser/expression/pratt/postfix.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            Some(Expr::Apply {
+                callee: Box::new(lhs),
+                args,
+            })
+        }
+    }
+
+    fn parse_bit_slice_postfix(&mut self, lhs: Expr) -> Option<Expr> {
+        self.ts.next_tok();
+        self.with_struct_literals_suspended(move |this| {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/parser/expression/pratt/postfix.rs:65]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | New modules (delay.rs, diff.rs, postfix.rs, classification.rs) contain zero internal unit tests and the refactored fixture suite omits diff/delay postfix test cases. | Add diff and delay postfix cases to src/parser/tests/expression/fixtures/valid/postfix.rs and implement targeted unit tests within delay.rs and diff.rs for handle\_diff\_marker, validate\_diff\_not\_pending, and parse\_delay\_postfix functions. |
| Module-Level Documentation | ❌ Error | Multiple new modules lack adequate module-level documentation. The pratt parser submodules (delay.rs, diff.rs, postfix.rs) contain no docstrings, and test fixture modules similarly lack module-level documentation explaining their purpose. | Add comprehensive module-level docstrings using //! to all newly created modules, documenting their purpose, responsibility, and integration points within the codebase. |
| Unit Architecture | ❌ Error | Three architectural violations identified: unhandled `next_tok()` side-effects in postfix.rs, overly strict early validation in diff.rs, and mutation hidden behind query-like naming in `validate_diff_not_pending`. | Bind three `next_tok()` calls to `_`; loosen diff.rs early guard; rename `validate_diff_not_pending` to clarify command semantics and document state mutations. |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 51.32% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Developer Documentation | ⚠️ Warning | PR introduces significant architectural changes and exposes new internal APIs but `docs/developers-guide.md` does not exist, and `docs/roadmap.md` does not mark issue `#223` as completed. | Create `docs/developers-guide.md` documenting the refactored parser architecture, new modules, and public APIs. Update `docs/roadmap.md` to mark issue `#223` as completed. |
| Testing (Unit And Behavioural) | ⚠️ Warning | New modules (sexpr.rs, classification.rs, delay.rs, diff.rs, postfix.rs) lack direct unit tests; test fixtures omit diff-marker and delay postfix cases; classification module has no dedicated unit tests for parse\_rule\_body\_term function. | Add inline unit tests to each new module covering local behaviour and edge cases. Add postfix fixture test cases for diff-marker ('expr\\'') and delay (-<) forms with expected Expr::AtomDiff and Expr::AtomDelay values. Add classification.rs unit tests for assignment parsing, for-loop classification, and error paths. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | The PR introduces Expr::to\_sexpr() with hardcoded assertion tests rather than snapshot testing infrastructure as required by issue `#261`, and lacks test fixtures for diff/delay postfix parsing functionality. | Integrate insta crate as dev-dependency, convert Expr::to\_sexpr() tests to snapshot-based assertions, and add explicit test fixtures for diff and delay postfix parsing to src/parser/tests/expression/fixtures/valid/postfix.rs. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/parser/tests/expression/fixtures/valid/postfix.rs](https://github.com/leynos/ddlint/pull/259/files/27ec6f66d794dca7b0bfd4287b1d980b90dde642#diff-ea272bc2a190c4f9cb7c8c1b3888728cb7b97ba063405c3d05421b19d5fd922f)

Comment on lines +10 to +101

~~~rust
pub(super) fn postfix_expression_cases() -> Vec<ExpressionCase> {
    vec![
        ExpressionCase {
            src: "(f)(x)",
            expected: call_expr(Expr::Group(Box::new(var("f"))), vec![var("x")]),
        },
        ExpressionCase {
            src: "foo.bar(x)",
            expected: method_call(var("foo"), "bar", vec![var("x")]),
        },
        ExpressionCase {
            src: "foo.bar(x).baz",
            expected: field_access(method_call(var("foo"), "bar", vec![var("x")]), "baz"),
        },
        ExpressionCase {
            src: "foo.bar",
            expected: field_access(var("foo"), "bar"),
        },
        ExpressionCase {
            src: "foo.bar.baz(x)",
            expected: method_call(field_access(var("foo"), "bar"), "baz", vec![var("x")]),
        },
        ExpressionCase {
            src: "foo.bar.baz().qux",
            expected: field_access(
                method_call(field_access(var("foo"), "bar"), "baz", vec![]),
                "qux",
            ),
        },
        ExpressionCase {
            src: "foo.bar().baz(x)",
            expected: method_call(
                method_call(var("foo"), "bar", vec![]),
                "baz",
                vec![var("x")],
            ),
        },
        ExpressionCase {
            src: "foo.bar.baz",
            expected: field_access(field_access(var("foo"), "bar"), "baz"),
        },
        ExpressionCase {
            src: "foo.bar().baz.qux(x)",
            expected: method_call(
                field_access(method_call(var("foo"), "bar", vec![]), "baz"),
                "qux",
                vec![var("x")],
            ),
        },
        ExpressionCase {
            src: "e[1,0]",
            expected: bit_slice(var("e"), lit_num("1"), lit_num("0")),
        },
        ExpressionCase {
            src: "t.0",
            expected: tuple_index(var("t"), "0"),
        },
        // Diff marker: e'(x) — diff wraps the completed call
        ExpressionCase {
            src: "e'(x)",
            expected: atom_diff(call_expr(var("e"), vec![var("x")])),
        },
        // Diff marker with bit-slice: e'[1,0]
        ExpressionCase {
            src: "e'[1,0]",
            expected: atom_diff(bit_slice(var("e"), lit_num("1"), lit_num("0"))),
        },
        // Delay postfix: e-<5>
        ExpressionCase {
            src: "e-<5>",
            expected: atom_delay(5, var("e")),
        },
        // Delay postfix with zero: e-<0>
        ExpressionCase {
            src: "e-<0>",
            expected: atom_delay(0, var("e")),
        },
        ExpressionCase {
            src: "S'(x)",
            expected: Expr::AtomDiff {
                expr: Box::new(call_expr(var("S"), vec![var("x")])),
            },
        },
        ExpressionCase {
            src: "A() -<10>",
            expected: Expr::AtomDelay {
                delay: 10,
                expr: Box::new(call_expr(var("A"), vec![])),
            },
        },
    ]
}
~~~

❌ New issue: [Large Method](https://codescene.io/projects/69278/delta?repo-id=1160463&review-id=259&biomarker=Large+Method&filename=src%2Fparser%2Ftests%2Fexpression%2Ffixtures%2Fvalid%2Fpostfix.rs&method=postfix_expression_cases)
postfix_expression_cases has 88 lines, threshold = 70

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following: **Inline unit tests for new modules** (`sexpr.rs`, `classification.rs`, `delay.rs`, `diff.rs`, `postfix.rs`) have no tracking issue. Shall I open one now?

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Large Method
src/parser/tests/expression/fixtures/valid/postfix.rs: postfix_expression_cases

### What lead to degradation?
postfix_expression_cases has 88 lines, threshold = 70

### Why does this problem occur?
Overly long functions make the code harder to read. The recommended maximum function length for the Rust language is 70 lines of code. Severity: Brain Method - Complex Method - Long Method.

### How to fix it?
We recommend to be careful here -- just splitting long functions don't necessarily make the code easier to read. Instead, look for natural chunks inside the functions that expresses a specific task or concern. Often, such concerns are indicated by a Code Comment followed by an if-statement. Use the [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html) refactoring to encapsulate that concern.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review
